[Feature] 공연 예매 API 구현

yeGenieee opened the pull request:
## 관련된 이슈 번호
- issue : #75 

## 변경사항 (할 일)
- [x] 공연 예매 API 구현
    - [x] 이미 공연 - 공연 시각 - 좌석이 매핑된 하나의 엔티티가 있다고 가정하고, 공연 예매를 할 수 있는 API를 구현합니다 


## 추가 점검사항
- [ ] test case
- [ ] document

yeGenieee committed:
공연 - 시각 - 좌석이 매핑된 하나의 Entity인 ScheduledShowSeat 추가

yeGenieee committed:
예매 API 스켈레톤만 구현

yeGenieee committed:
좌석 예매 API 구현

yeGenieee committed:
매핑된 공연 - 시각 - 좌석 테이블 내 공연 시각 조회 파라미터 타입을 String -> LocalDateTime 으로 변경

yeGenieee committed:
checkStyle 맞춤

yeGenieee committed:
dummy_data insert문 추가

yeGenieee committed:
init_ddl.sql 내 N/N 오타 수정

yeGenieee committed:
init_ddl.sql 내 AUTO_INCREMENT 누락 사항 추가

yeGenieee committed:
UserRepository 내 find메소드 명칭 구체화

yeGenieee committed:
init_ddl.sql 내 drop table 구문 추가 및 insert_dummy_data dateformat 형식 수정

yeGenieee committed:
Revert commit for 'UserRepository 내 find메소드 명칭 구체화'

yeGenieee committed:
예매 Request 내 좌석 count만큼 예매 데이터가 생성되도록 (예매하도록) 구현

yeGenieee committed:
DateTimeUtils 제거

yeGenieee committed:
Interceptor 테스트 중 잘못된 URL 수정

pollra commented on app/src/main/java/com/picketing/www/business/domain/reservation/ReservationFactory.java:
```diff
@@ -0,0 +1,38 @@
+			.build();
+	}
+
+	public ReservationIdResponse convertReservationId(Reservation reservation) {
```
해당 factory 함수는 내부에서만 사용 될 목적으로 만들어진 것 같아보여요
함수를 하나로 합치거나 private 으로 바꾸는것은 어떨까요?

yeGenieee replied:
`private` 으로 변경하겠습니다

pollra commented on app/src/main/java/com/picketing/www/business/service/reservation/ReservationService.java:
```diff
@@ -0,0 +1,82 @@
+		String showTime = request.showTime();
+
+		// 각 좌석이 구매 가능한지 확인
+		if (!checkAvailable(show, showTime, request.seatGradeList())) {
```
check 는 확인만 한다는 느낌이라서 리턴되는 boolean 값을 예측하기 힘들어보여요
isAvailableShow 같은 네이밍은 어떻게 생각하시나요?

yeGenieee replied:
`isBookable` 로 변경하면 좋을 것 같습니다

pollra commented on app/src/main/java/com/picketing/www/persistence/repository/show/ShowRepository.java:
```diff
@@ -1,19 +1,15 @@
+	Optional<Show> findById(Long showId);
 
-	List<Show> findShowsByGenreAndSubGenre(Genre genre, SubGenre subGenre, Pageable pageable);
+	// List<Show> findShowsByGenreAndSubGenre(Genre genre, SubGenre subGenre, Pageable pageable);
```
주석이 남겨진 이유가 궁금해요

yeGenieee replied:
아 ! 이건 해당 로직을 이전에 PR에서 구현하다가 말았던 부분이라서, 다 완성된 것이 아니라고 생각해서 주석 처리해두었습니다
아래에서 비슷하게 의견주신 것과 같이 `TODO`로 명시해두겠습니다

pollra commented on app/src/main/java/com/picketing/www/presentation/dto/request/reservation/ReservationRequest.java:
```diff
@@ -0,0 +1,25 @@
+
+	@Builder
+	public record ReservationSeatRequest(
+		String seatGrade,
```
SeatGrade 가 enum 으로 변경되어서, enum 에 대한 검증 로직이 없다면 valueOf 로 변환 과정에서 문제가 생길 수 있을 것 같아보여요
Enum 으로 바로 받아서 처리하는건 어떻게 생각하시나요?

yeGenieee replied:
넵 현재는 `SeatGrade`가 enum이므로,  Enum으로 바로 받아서 처리하는 게 더 좋을 것 같습니다 변경하겠습니다

pollra commented on app/src/main/java/com/picketing/www/business/domain/reservation/ReservationFactory.java:
```diff
@@ -0,0 +1,38 @@
+		return ReservationIdResponse.builder()
+			.id(reservation.getId())
+			.build();
+	}
```
개인적으로는 builder 를 사용 할 정도의 복잡성이 느껴지지는 않아보여요 (필드가 하나니까요)
new Reservation(id) 로 하면 함수로 뺄 필요가 없을 것 같은데 어떻게 생각하시나요

pollra commented on app/src/main/java/com/picketing/www/business/service/show/ShowService.java:
```diff
@@ -23,7 +25,8 @@ public class ShowService {
 
 	public List<Show> getShowList(Genre genre, SubGenre subGenre, Pageable pageable) {
-		return showRepository.findShowsByGenreAndSubGenre(genre, subGenre, pageable);
+		// return showRepository.findShowsByGenreAndSubGenre(genre, subGenre, pageable);
```
추후에 처리 될 주석일까요?
ShowService 에서 이러한 주석이 몇개 보이는데 어떻게 처리될 예정인지 적혀있지 않아서 의도 한 것인지 구분이 어려워요
todo 같은 주석이 있으면 더 좋을 것 같아요

pollra commented on app/src/main/java/com/picketing/www/business/service/reservation/ReservationService.java:
```diff
@@ -0,0 +1,82 @@
+
+	private boolean checkAvailable(Show show, String showTime, List<ReservationSeatRequest> seatRequestList) {
+		return seatRequestList.stream()
+			.allMatch(s -> {
```
s 보다는 seatRequest 같은 네이밍이 좀 더 어울릴 것 같은데 어떻게 생각하시나요?

yeGenieee replied:
로컬 변수라서 크게 신경쓰지 않았는데, 수정하겠습니다

pollra commented on app/src/main/java/com/picketing/www/presentation/dto/request/reservation/ReservationRequest.java:
```diff
@@ -0,0 +1,25 @@
+
+	String showTime,
+
+	List<ReservationSeatRequest> seatGradeList
```
요청 값에대한 검증이 필요해보여요

pollra commented:
코드리뷰 내용도 있고 궁금한 부분도 있지만, 급한 사항이라고 생각해서 일단은 approve 처리했습니답

yeGenieee committed:
코드 리뷰 사항 반영
1. 접근제한자 수정 (public -> private)
2. 네이밍 적절하게 변경 (checkAvailable -> isBookable)
3. TODO 주석 추가
4. Request 내 Enum 필드로 바로 받도록 수정

yeGenieee committed:
validation dependency 추가

f-lab-lion commented on app/src/main/java/com/picketing/www/business/service/reservation/ReservationService.java:
```diff
@@ -0,0 +1,82 @@
+
+		return reservationRepository.saveAll(reservations);
+	}
+
```
```java
        public List<Reservation> makeReservations(ReservationRequest request) {
        	User user = userService.get(request.userId());

		Show show = showService.getShowById(request.showId());

		String showTime = request.showTime();
		
		
		List<Reservation> reservations = request.seatGradeList()
			.stream()
			.flatMap(seatRequest -> makeReservationPerCount(user, show, showTime, seatRequest).stream())
			.collect(Collectors.toList());
			
		makeReservations(user, show, showTime, reservations)

        }
        
	public List<Reservation> makeReservations(User user,Show show, String showTime, List<Reservation> reservations ) {

		// 각 좌석이 구매 가능한지 확인
		if (!isBookable(show, showTime, reservations) {
			throw new CustomException(ErrorCode.ALREADY_RESERVED);
		}

		List<Reservation> reservations = request.seatGradeList()
			.stream()
			.flatMap(seatRequest -> makeReservationPerCount(user, show, showTime, seatRequest).stream())
			.collect(Collectors.toList());

		return reservationRepository.saveAll(reservations);
	}
```
